std/utf.d

```diff
@@ -1132,6 +1132,150 @@ dchar decodeFront(UseReplacementDchar useReplacementDchar = No.useReplacementDch
     return decodeFront!useReplacementDchar(str, numCodeUnits);
 }
 
+/++
+    $(D decodeBack) is a variant of $(LREF decode) which specifically decodes
+    the last code point. Unlike $(LREF decode), $(D decodeBack) accepts any
+    bidirectional range of code units (rather than just a string or random access
+    range). It also takes the range by $(D ref) and pops off the elements as it
+    decodes them. If $(D numCodeUnits) is passed in, it gets set to the number
+    of code units which were in the code point which was decoded.
+
+    Params:
+        useReplacementDchar = if invalid UTF, return replacementDchar rather than throwing
+        str = input string or bidirectional Range
+        numCodeUnits = set to number of code units processed
+
+    Returns:
+        decoded character
+
+    Throws:
+        $(LREF UTFException) if $(D str.back) is not the end of a valid UTF
+        sequence. If an exception is thrown, the $(D str) itself remains unchanged,
+        but there is no guarantee as to the value of $(D numCodeUnits) (when passed).
+  +/
+dchar decodeBack(UseReplacementDchar useReplacementDchar = No.useReplacementDchar, S)(
+    ref S str, out size_t numCodeUnits) @trusted pure if (isSomeString!S)
```

Comment:
Applying `@trusted` to an entire function is dangerous. If someone adds in unsafe code at a later date, then the function is still marked trusted with out a warning to the compiler. What we do instead is something called a trusted escape, where the code to be trusted is put in an anonymous closure marked `@trusted` and then immediately called. E.g.

```d
() @trusted { my_unsafe_func(); } ()
```

Also, `pure` should be inferred rather than specified.